diff --git a/tensorflow/core/kernels/conv_grad_ops_3d.cc b/tensorflow/core/kernels/conv_grad_ops_3d.cc
index 08e832760924e6..cc6e3ffd90075c 100644
--- a/tensorflow/core/kernels/conv_grad_ops_3d.cc
+++ b/tensorflow/core/kernels/conv_grad_ops_3d.cc
@@ -741,8 +741,6 @@ class Conv3DBackpropFilterOp : public OpKernel {
       filter_shape = context->input(1).shape();
     }
 
-<<<<<<< HEAD
-=======
     OP_REQUIRES(context, input_shape.dims() == 5,
                 errors::InvalidArgument("input tensor must have 5 dimensions"));
     OP_REQUIRES(
@@ -765,7 +763,6 @@ class Conv3DBackpropFilterOp : public OpKernel {
                                 " for out_backprop and ",
                                 filter_shape.dim_size(4), " for filter_sizes"));
 
->>>>>>> 311403edbc9 (Eliminate a division by 0 in 3D convolutions.)
     ConvBackpropDimensions dims;
     OP_REQUIRES_OK(context,
                    ConvBackpropComputeDimensions(
@@ -878,8 +875,6 @@ class Conv3DCustomBackpropFilterOp : public OpKernel {
       filter_shape = context->input(1).shape();
     }
 
-<<<<<<< HEAD
-=======
     OP_REQUIRES(context, input_shape.dims() == 5,
                 errors::InvalidArgument("input tensor must have 5 dimensions"));
     OP_REQUIRES(
@@ -902,7 +897,6 @@ class Conv3DCustomBackpropFilterOp : public OpKernel {
                                 " for out_backprop and ",
                                 filter_shape.dim_size(4), " for filter_sizes"));
 
->>>>>>> 311403edbc9 (Eliminate a division by 0 in 3D convolutions.)
     ConvBackpropDimensions dims;
     OP_REQUIRES_OK(context,
                    ConvBackpropComputeDimensions(
