diff --git a/tensorflow/core/grappler/optimizers/constant_folding.cc b/tensorflow/core/grappler/optimizers/constant_folding.cc
index df4cc54757..9dc97639ce 100644
--- a/tensorflow/core/grappler/optimizers/constant_folding.cc
+++ b/tensorflow/core/grappler/optimizers/constant_folding.cc
@@ -1670,7 +1670,9 @@ bool ConstantFolding::IsSimplifiableReshape(
   if (!IsReshape(node)) {
     return false;
   }
-  CHECK_LE(2, node.input_size());
+  if (2 > node.input_size()) {
+    return false;
+  }
   const NodeDef* new_shape = node_map_->GetNode(node.input(1));
   if (!IsReallyConstant(*new_shape)) {
     return false;
@@ -1686,7 +1688,9 @@ bool ConstantFolding::IsSimplifiableReshape(
   if (!s.ok()) {
     return false;
   }
-  CHECK_EQ(1, outputs.size());
+  if (outputs.size() != 1) {
+    return false;
+  }
 
   const std::vector<OpInfo::TensorProperties>& props =
       properties.GetInputProperties(node.name());
