diff --git a/tensorflow/lite/kernels/conv.cc b/tensorflow/lite/kernels/conv.cc
index 4909007562..ad33a9fd9b 100644
--- a/tensorflow/lite/kernels/conv.cc
+++ b/tensorflow/lite/kernels/conv.cc
@@ -427,6 +427,7 @@ TfLiteStatus Prepare(KernelType kernel_type, TfLiteContext* context,
     // Only one scale factor per batch is typically necessary. See optimized
     // implementation for why we need to allocate for the height of the inputs
     // flattened to 2D.
+    TF_LITE_ENSURE(context, channels_in != 0);
     const int height = NumElements(input) / channels_in;
     int scaling_dims[1] = {height};
     if (!TfLiteIntArrayEqualsArray(scaling_factors->dims, 1, scaling_dims)) {
@@ -449,6 +450,7 @@ TfLiteStatus Prepare(KernelType kernel_type, TfLiteContext* context,
       input_offsets->type = kTfLiteInt32;
       input_offsets->allocation_type = kTfLiteArenaRw;
       // See above comment for the need to allocate for height of inputs.
+      TF_LITE_ENSURE(context, channels_in != 0);
       const int height = NumElements(input) / channels_in;
       int scaling_dims[1] = {height};
       if (!TfLiteIntArrayEqualsArray(input_offsets->dims, 1, scaling_dims)) {
@@ -647,17 +649,19 @@ void EvalFloat(TfLiteContext* context, TfLiteNode* node,
 }
 
 template <KernelType kernel_type>
-void EvalHybridPerChannel(TfLiteContext* context, TfLiteNode* node,
-                          TfLiteConvParams* params, OpData* data,
-                          TfLiteTensor* input, TfLiteTensor* filter,
-                          TfLiteTensor* bias, TfLiteTensor* im2col,
-                          TfLiteTensor* hwcn_weights, TfLiteTensor* output) {
+TfLiteStatus EvalHybridPerChannel(TfLiteContext* context, TfLiteNode* node,
+                                  TfLiteConvParams* params, OpData* data,
+                                  TfLiteTensor* input, TfLiteTensor* filter,
+                                  TfLiteTensor* bias, TfLiteTensor* im2col,
+                                  TfLiteTensor* hwcn_weights,
+                                  TfLiteTensor* output) {
   float output_activation_min, output_activation_max;
   CalculateActivationRange(params->activation, &output_activation_min,
                            &output_activation_max);
 
-  const int input_size = NumElements(input) / SizeOfDimension(input, 0);
   const int batch_size = SizeOfDimension(input, 0);
+  TF_LITE_ENSURE(context, batch_size != 0);
+  const int input_size = NumElements(input) / batch_size;
   const TfLiteTensor* input_quantized =
       GetTemporary(context, node, data->input_quantized_index);
   int8_t* quantized_input_ptr_batch = input_quantized->data.int8;
@@ -715,19 +719,22 @@ void EvalHybridPerChannel(TfLiteContext* context, TfLiteNode* node,
       break;
     }
   }
+  return kTfLiteOk;
 }
 
 template <KernelType kernel_type>
-void EvalHybrid(TfLiteContext* context, TfLiteNode* node,
-                TfLiteConvParams* params, OpData* data, TfLiteTensor* input,
-                TfLiteTensor* filter, TfLiteTensor* bias, TfLiteTensor* im2col,
-                TfLiteTensor* hwcn_weights, TfLiteTensor* output) {
+TfLiteStatus EvalHybrid(TfLiteContext* context, TfLiteNode* node,
+                        TfLiteConvParams* params, OpData* data,
+                        TfLiteTensor* input, TfLiteTensor* filter,
+                        TfLiteTensor* bias, TfLiteTensor* im2col,
+                        TfLiteTensor* hwcn_weights, TfLiteTensor* output) {
   float output_activation_min, output_activation_max;
   CalculateActivationRange(params->activation, &output_activation_min,
                            &output_activation_max);
 
-  const int input_size = NumElements(input) / SizeOfDimension(input, 0);
   const int batch_size = SizeOfDimension(input, 0);
+  TF_LITE_ENSURE(context, batch_size != 0);
+  const int input_size = NumElements(input) / batch_size;
 
   float* input_ptr = GetTensorData<float>(input);
   int8_t* quantized_input_ptr_batch = GetTensorData<int8_t>(
@@ -772,6 +779,7 @@ void EvalHybrid(TfLiteContext* context, TfLiteNode* node,
       break;
     }
   }
+  return kTfLiteOk;
 }
 
 template <KernelType kernel_type>
@@ -807,12 +815,13 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
     case kTfLiteFloat32:
       if (filter->type == kTfLiteUInt8 || filter->type == kTfLiteInt8) {
         if (is_hybrid_per_channel) {
-          EvalHybridPerChannel<kernel_type>(context, node, params, data, input,
-                                            filter, bias, im2col, hwcn_weights,
-                                            output);
+          TF_LITE_ENSURE_STATUS(EvalHybridPerChannel<kernel_type>(
+              context, node, params, data, input, filter, bias, im2col,
+              hwcn_weights, output));
         } else {
-          EvalHybrid<kernel_type>(context, node, params, data, input, filter,
-                                  bias, im2col, hwcn_weights, output);
+          TF_LITE_ENSURE_STATUS(EvalHybrid<kernel_type>(
+              context, node, params, data, input, filter, bias, im2col,
+              hwcn_weights, output));
         }
       } else {
         EvalFloat<kernel_type>(context, node, params, data, input, filter, bias,
